Added Study Room Functions to to Library Module

ghosteau opened the pull request:
- Added a function to check reserved Hillman library study room times, tied to which room is specifically reserved
- Added another function to check the total amount of reserved rooms in the Hillman library

ghosteau committed:
Added a couple Hillman library features

Tells the user how many reservations are in the Hillman library, and also tells the user what rooms and times are reserved in the Hillman library.

ghosteau committed:
fixed a bug with reservations pulling; moved reservations functions to library module

ghosteau committed:
revert lab back

ghosteau committed:
revert lab

ghosteau committed:
Update lab.py

ghosteau committed:
revert lab again

ghosteau reviewed (commented):
Some  very slight spacing changes may have been made to lab module because I wanted to revert it back on my fork. Just ignore all changes outside of the library module

tianyizheng02 committed:
Merge branch 'dev' into in_prog

tianyizheng02 commented:
Please run `black .` in your local git repo to fix the formatting issues

ghosteau committed:
Ran black to format correctly

ghosteau committed:
refromatted with black

ghosteau commented:
I ran black through cmd, let me know if there is anything else that needs done.

tianyizheng02 commented on pittapi/course.py:
```diff
@@ -215,7 +215,16 @@ def get_course_details(term: Union[str, int], subject: str, course: Union[str, i
         )
 
     return CourseDetails(
+<<<<<<< HEAD
+        course=Course(
+            subject_code=subject,
+            course_number=course,
+            course_id=internal_course_id,
+            course_title=course_title,
+        ),
+=======
         course=Course(subject_code=subject, course_number=course, course_id=internal_course_id, course_title=course_title),
+>>>>>>> 5707e24e2f881755bb429b3b1a0f2245940b46d6
```
Some of the files have merge conflicts that weren't properly resolved and need to be cleaned up, e.g.
```suggestion
        course=Course(
            subject_code=subject,
            course_number=course,
            course_id=internal_course_id,
            course_title=course_title,
        ),
```
The stuff between the `<<<` and `===` is code from upstream and the stuff between the `===` and `>>>` is code from your local git repo, and this is a merge conflict because the two versions differ too much for git to automatically merge your changes in. Please look in the other modified files for anything like this and resolve those conflicts (i.e., choose one version or the other and clean up any labels that git added).

ghosteau committed:
Update pittapi/course.py

Co-authored-by: Tianyi Zheng <tianyizheng02@gmail.com>

tianyizheng02 commented on black:
?

ghosteau replied:
Working on it now. Where do I find these merge conflicts in my IDE so I can delete the problematic code?

tianyizheng02 replied:
IDK if they'll necessarily show up on your IDE because you already committed the code to git. If they do show up, tho, they'll be syntax errors since they're definitely not valid (Python) code. I recommend you just go thru the files under the Files Changed tab of this PR and take a close look at your changes.

Also, the "?" comment is about this file: what's it doing there?

ghosteau replied:
Oh, no idea how that file got there. Probably somehow got there when I ran the black command; should be deleted now. Thanks for the help, I am gonna sift through the duplicates and merge error code now.

ghosteau committed:
Delete black

ghosteau committed:
Update course_test.py

ghosteau committed:
Update dining_test.py

ghosteau committed:
Update library_test.py

ghosteau committed:
Update course_mocks.py

ghosteau committed:
Update people_test.py

ghosteau committed:
Update sports_test.py

ghosteau committed:
Update status_test.py

ghosteau committed:
Update textbook_test.py

ghosteau committed:
Update textbook_test.py

ghosteau commented:
I think I dealt with the merger conflicts -- I elected to choose the segments that were formatted with black to hopefully make readability for the API better for some portions. Do let me know if there is anything else, or anything I missed.

tianyizheng02 commented on tests/course_test.py:
```diff
@@ -130,15 +127,11 @@ def test_get_course_details(self):
             test_attribute.attribute_description, "*DSAS General Ed. Requirements"
         )
-=======
         self.assertEqual(test_attribute.attribute_description, "*DSAS General Ed. Requirements")
```
This merge conflict was resolved by including both versions rather than just one

ghosteau replied:
Not sure I entirely did what you asked with this one, so do let me know if I screwed it up on my end. I got rid of the bottom line because it was redundant, and it's likely I forgot to delete it when cleaning up duplicates.

tianyizheng02 replied:
Yup, that's precisely what I was referring to. The problem was that one version of the code had
```py
self.assertEqual(
    test_attribute.attribute_description, "*DSAS General Ed. Requirements"
)
```
while the other version of the code had
```py
self.assertEqual(test_attribute.attribute_description, "*DSAS General Ed. Requirements")
```
and git was showing both versions for you to pick one of the two, so one of them needed cleaning up.

tianyizheng02 commented on tests/dining_test.py:
```diff
@@ -85,11 +85,6 @@ def test_get_location_menu(self):
-=======
-        locations = dining.get_location_menu("The Eatery", datetime.datetime(2024, 4, 12), "Breakfast")
->>>>>>> 5707e24e2f881755bb429b3b1a0f2245940b46d6
-        self.assertIsInstance(locations, dict)
```
This line was accidentally deleted when you cleaned up this merge conflict, but this line isn't part of the merge conflict

tianyizheng02 commented on tests/mocks/course_mocks.py:
```diff
@@ -1,17 +1,12 @@
         }
     ]
-=======
     "courses": [{"catalog_nbr": "0007", "descr": "INTRODUCTION TO COMPUTER PROGRAMMING", "crse_id": "105611"}]
```
This merge conflict was resolved by including both versions rather than just one

tianyizheng02 commented on tests/mocks/course_mocks.py:
```diff
@@ -1,17 +1,12 @@
-mocked_subject_data = {"subjects": [{"subject": "CS", "descr": "Computer Science"}]}
-
```
These lines were accidentally deleted when you cleaned up this merge conflict, but these lines aren't part of the merge conflict

tianyizheng02 commented on tests/textbook_test.py:
```diff
@@ -269,9 +220,4 @@ def test_invalid_department_code(self):
             ValueError, textbook.get_textbook, TERM, "TEST", "000", "EXIST", None
         )
-=======
         self.assertEqual(textbook._filter_dictionary(test_dict, test_key), {"a": 1, "c": 3})
```
This line should've been deleted when you cleaned up this merge conflict, since it's part of the version that you deleted

ghosteau replied:
I deleted that last line  "self.assertEqual(textbook._filter_dictionary(test_dict, test_key), {"a": 1, "c": 3})" if that is what you were referring to.

ghosteau committed:
Update course_mocks.py

ghosteau committed:
Update dining_test.py

ghosteau committed:
Update course_test.py

ghosteau committed:
Update textbook_test.py

ghosteau committed:
Update course_mocks.py

tianyizheng02 reviewed (commented):
I think all the biggest issues (syntax errors, etc) have been fixed, so we can now let GitHub run our repo workflows on this PR before I comment on anything else in the code

tianyizheng02 committed:
Fix black errors

tianyizheng02 commented:
So there were still more errors with `black` cuz we run `black` with extra command-line args and that was messing with things... I ran `black` again locally and pushed a commit to get the workflows to pass, sorry about all the fuss

tianyizheng02 commented on pittapi/library.py:
```diff
@@ -127,3 +130,38 @@ def _extract_facets(facet_fields: List[Dict[str, Any]]) -> Dict[str, List[Dict[s
     return facets
+
+
+def hillman_total_reserved():
```
Could you add a type hint for the return type of this function?

tianyizheng02 commented on pittapi/library.py:
```diff
@@ -127,3 +130,38 @@ def _extract_facets(facet_fields: List[Dict[str, Any]]) -> Dict[str, List[Dict[s
+    return count
+
+
+def reserved_hillman_times():
```
Same with the return type of this function

tianyizheng02 commented on pittapi/library.py:
```diff
@@ -127,3 +130,38 @@ def _extract_facets(facet_fields: List[Dict[str, Any]]) -> Dict[str, List[Dict[s
+
+def reserved_hillman_times():
+    """Returns a list of dictionaries of reserved rooms of the Hillman with their respective times"""
+    bookings = []
+
+    resp = requests.get(STUDY_ROOMS_URL)
+    resp = resp.json()
+    data = resp["data"]
+
+    if data is None:
+        return bookings
+
+    # Note: there can be multiple reservations in the same room, hence why we must use a list of maps, and cannot just use a singular map
+    for reservation in data:
+        bookings.append(
+            {
+                "Room": reservation["itemName"],
+                "Reserved": [reservation["from"], reservation["to"]],
+            }
+        )
+    return bookings
```
```suggestion
    resp = requests.get(STUDY_ROOMS_URL)
    resp = resp.json()
    data = resp["data"]

    if data is None:
        return []

    # Note: there can be multiple reservations in the same room, hence why we must use a list of maps, and cannot just use a singular map
    bookings = [
        {
            "Room": reservation["itemName"],
            "Reserved": [reservation["from"], reservation["to"]],
        }
        for reservation in data
    ]
    return bookings
```
The for-loop at the end here can be written as a list comprehension. Most of the time, it's more readable and more efficient in Python to use a comprehension rather than manually iterating over a collection and appending things.

tianyizheng02 commented on pittapi/library.py:
Could we also get unit tests for these new functions that you're adding?

ghosteau replied:
Yeah, I've never used it so I'll have to look at the docs when I get home but I can definitely work on that. Thanks for the help!

ghosteau committed:
Update library.py

Co-authored-by: Tianyi Zheng <tianyizheng02@gmail.com>

ghosteau committed:
Added type hints for new functions

tianyizheng02 commented on pittapi/library.py:
```diff
@@ -162,4 +161,4 @@ def reserved_hillman_times():
         for reservation in data
     ]
-    return bookings
+    return bookings
```
Please make sure that all files end with an empty line (we check for this as part of our pre-commit)

tianyizheng02 commented on pittapi/library.py:
```diff
@@ -144,8 +144,7 @@ def hillman_total_reserved():
 
-
-def reserved_hillman_times():
+def reserved_hillman_times() -> List[Dict[str, Union[str, List[str]]]]:
```
In Python, the style convention is that there should be two empty lines between functions (unless you're in a class, in which case there should be one empty line between methods)

tianyizheng02 commented on pittapi/library.py:
```diff
@@ -19,7 +19,7 @@
 import requests
 from html.parser import HTMLParser
-from typing import Any, Dict, List
+from typing import Any, Dict, List, Union
```
FYI, Python has built-in type hints for the most common types such as `list`, `dict`, etc., which means that you don't need most of these imports from the `typing` library. For example, instead of
```py
from typing import Dict, List, Union

...

def reserved_hillman_times() -> List[Dict[str, Union[str, List[str]]]]:
```
you can now write
```py
...

def reserved_hillman_times() -> list[dict[str, str | list[str]]]:
```
I'm not gonna ask you to change all the type hints to not use the `typing` library (since this file was already using it), but it's good to know that these types are now a built-in feature.

ghosteau replied:
Alright. I didn't realize this was natively supported. Good to know for future reference.

tianyizheng02 reviewed (changes requested):
The GitHub workflow is failing because of formatting issues. For future contributions, I highly, highly recommend that you set up pre-commit and other tooling that we use in this repo (I have a write-up on how to set up everything in #170, but it hasn't been merged yet). Pre-commit automatically checks for certain issues and runs tools like `black` _before_ you make a git commit, so it streamlines everything and you don't need to worry about formatting and stuff yourself.

ghosteau committed:
Polished up formatting in library file

- Added space at end of file
- Enforced 2 spaces between functions

ghosteau committed:
Merge branch 'dev' into in_prog

ghosteau committed:
Added unit tests for library module update

I added unit tests for both of the new functions and updated the sample path variable.

For some reason I was having issues with the original sample path, so I updated it so the testing works good for all functions in the library_test file now.

ghosteau committed:
Merge branch 'in_prog' of https://github.com/ghosteau/PittAPI into in_prog

ghosteau committed:
changed spacing for a line

ghosteau commented:
I wrote the unit tests and I believe I cleaned up all of the slight errors with the hints and spacing. Let me know if there is anything else that needs done!

For the unit tests I did make a new class for the new extension of the library module. If you want, I can add it to the old class, but I figured it would be easier to separate them since they do use different links and different mock JSONs.

I also had to update the path to be `SAMPLE_PATH = Path(__file__).parent / "samples"` instead, because for some reason the tests were encountering issues using the previous method.

tianyizheng02 commented:
I think it looks good, let's run the test workflow and see if there are any issues

tianyizheng02 commented:
More black errors, this time with some minor indentation stuff (though personally I think your code looks fine). Could you run `black --line-length=127 .`?

tianyizheng02 commented on tests/library_test.py:
```diff
@@ -36,8 +36,58 @@ def __init__(self, *args, **kwargs):
+
+    @responses.activate
+    def test_hillman_total_reserved(self):
+        responses.add(
+            responses.GET,
+            library.STUDY_ROOMS_URL,
+            json=self.hillman_query,
+            status=200
+        )
```
```suggestion
        responses.add(
            responses.GET,
            library.STUDY_ROOMS_URL,
            json=self.hillman_query,
            status=200,
        )
```
`black` wants to collapse this down into a single line, but I assume you intended to write the arguments on separate lines. If you add a trailing comma to the last argument, it tells `black` that you don't want the arguments collapsed into a single line.

ghosteau replied:
Yeah that was a small error on my part. I thought I might need an extra argument but I didn't.

tianyizheng02 commented on tests/library_test.py:
```diff
@@ -36,8 +36,58 @@ def __init__(self, *args, **kwargs):
+
+    @responses.activate
+    def test_reserved_hillman_times(self):
+        responses.add(
+            responses.GET,
+            library.STUDY_ROOMS_URL,
+            json=self.hillman_query,
+            status=200
+        )
```
```suggestion
        responses.add(
            responses.GET,
            library.STUDY_ROOMS_URL,
            json=self.hillman_query,
            status=200,
        )
```
Same thing here

tianyizheng02 reviewed (changes requested):
`black` is throwing an error again because of some minor formatting stuff (even though personally I think your code is fine)... see my comments below

tianyizheng02 commented on tests/library_test.py:
```diff
@@ -36,8 +36,58 @@ def __init__(self, *args, **kwargs):
+            json=self.hillman_query,
+            status=200
+        )
+        mock_answer = [
+            {
+                "Room": "408 HL (Max. 5 persons) (Enclosed Room)",
+                "Reserved": ["2024-06-12 17:30:00", "2024-06-12 20:30:00"]
+            },
+            {
+                "Room": "409 HL (Max. 5 persons) (Enclosed Room)",
+                "Reserved": ["2024-06-12 18:00:00", "2024-06-12 21:00:00"]
+            },
+            {
+                "Room": "303 HL (Max. 5 persons) (Enclosed Room)",
+                "Reserved": ["2024-06-12 18:30:00", "2024-06-12 21:30:00"]
+            },
+            {
+                "Room": "217 HL (Max. 10 persons) (Enclosed Room)",
+                "Reserved": ["2024-06-12 19:00:00", "2024-06-12 22:30:00"]
+            }
+        ]
```
```suggestion
        mock_answer = [
            {
                "Room": "408 HL (Max. 5 persons) (Enclosed Room)",
                "Reserved": ["2024-06-12 17:30:00", "2024-06-12 20:30:00"],
            },
            {
                "Room": "409 HL (Max. 5 persons) (Enclosed Room)",
                "Reserved": ["2024-06-12 18:00:00", "2024-06-12 21:00:00"],
            },
            {
                "Room": "303 HL (Max. 5 persons) (Enclosed Room)",
                "Reserved": ["2024-06-12 18:30:00", "2024-06-12 21:30:00"],
            },
            {
                "Room": "217 HL (Max. 10 persons) (Enclosed Room)",
                "Reserved": ["2024-06-12 19:00:00", "2024-06-12 22:30:00"],
            },
        ]
```
Same thing here

ghosteau committed:
Update tests/library_test.py

Co-authored-by: Tianyi Zheng <tianyizheng02@gmail.com>

ghosteau committed:
Update tests/library_test.py

Co-authored-by: Tianyi Zheng <tianyizheng02@gmail.com>

ghosteau committed:
Update tests/library_test.py

Co-authored-by: Tianyi Zheng <tianyizheng02@gmail.com>

tianyizheng02 commented:
Everything looks good! The only thing left is to update all the type hints to use the built-in types rather than the `typing` library types (I cleaned up most of the types throughout the repo in a different PR, so it's best that we keep things consistent now). I'll handle this edit myself because GitHub's suggestion system limits where I can leave suggestions.

tianyizheng02 committed:
Update type hints

ghosteau commented:
Sounds good! I appreciate all of the help with the formatting especially!

tianyizheng02 reviewed (approved):
All workflows/tests are passing, and I think everything looks good! Thanks for your work and for your patience with the tooling. As mentioned earlier, I highly recommend learning how to install and use pre-commit and other tools (see CONTRIBUTING.md), because that'll make the PR process much easier and more straightforward for everyone.
